video/out/drm_prime.c

```diff
@@ -78,15 +81,64 @@ int drm_prime_create_framebuffer(struct mp_log *log, int fd, AVDRMFrameDescripto
    return -1;
 }
 
-void drm_prime_destroy_framebuffer(struct mp_log *log, int fd, struct  drm_prime_framebuffer *framebuffer)
+void drm_prime_destroy_framebuffer(struct mp_log *log, int fd,
+                                    struct drm_prime_framebuffer *framebuffer,
+                                    struct drm_prime_handle_refs *handle_refs)
 {
     if (framebuffer->fb_id)
         drmModeRmFB(fd, framebuffer->fb_id);
 
     for (int i = 0; i < AV_DRM_MAX_PLANES; i++) {
-        if (framebuffer->gem_handles[i])
-            drmIoctl(fd, DRM_IOCTL_GEM_CLOSE, &framebuffer->gem_handles[i]);
+        if (framebuffer->gem_handles[i]) {
+            drm_prime_remove_handle_ref(handle_refs,
+                                         framebuffer->gem_handles[i]);
+            if(!drm_prime_get_handle_ref_count(handle_refs, framebuffer->gem_handles[i])) {
+                drmIoctl(fd, DRM_IOCTL_GEM_CLOSE, &framebuffer->gem_handles[i]);
+            }
+        }
     }
 
     memset(framebuffer, 0, sizeof(*framebuffer));
 }
+
+void drm_prime_init_handle_ref_count(void *talloc_parent, struct drm_prime_handle_refs *handle_refs) {
+    //4 dumb buffer + (1 per plane * planes we keep) + 10 (extra safety)
+    uint16_t init_size = 4 + (AV_DRM_MAX_PLANES * 4) + 10;
+    handle_refs->handle_ref_count = talloc_size(talloc_parent, init_size * sizeof(uint32_t));
+    handle_refs->size = init_size;
+    handle_refs->ctx = talloc_parent;
+
+    memset(handle_refs->handle_ref_count, 0, sizeof(uint32_t) * init_size);
+}
+
+void drm_prime_add_handle_ref(struct drm_prime_handle_refs *handle_refs, uint32_t handle) {
+    if(handle) {
+        if(handle > handle_refs->size) {
+            handle_refs->size *= 2;
```

Comment:
Not necessary, just set it to maximum+1 or so, or just change handle_refs.size to handle_refs.max (which would include the last item).

Reply:
Also, for clarification: the MP_TARRAY_GROW pre-allocates, so you don't need to provide preallocation logic.

Reply:
You mean that doubling the size is not necessary ? I don't know how talloc works but I wanted to not resize and copy that often. If it doubles the size we should be safe to not have to have a copy and resize again. Thats was my opinion why size is doubled not +1.

Reply:
Not having to do that is half the purpose of the talloc stuff: https://github.com/mpv-player/mpv/blob/master/ta/ta_talloc.h#L94

Reply:
Okay I got it. I wil change that so that it only resizes.
What I got from code is that it -1 and then doubles the size.